accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,185 @@
+    }
+```
+
+***<u>OpenQuestion</u>***
+
+We need to decide about the behavior of the old property `Activity.Tags` when adding tags of non-string types values. The options are
+
+- Ignore all non-string typed values and return only the string type value tags.
+- Convert the non-string type values to string before returning it.
+
+I am inclining to the first option to avoid handling ToString with more complex types (e.g. arrays or ActivityTagsCollection).
+
+## Proposal Changing already exposed APIs in previous preview releases
+
+As we are introducing the ActivityTagsCollection, we can take advantage of that now and modify some of the exposed APIs to use this collection instead. Also, we are renaming Attributes to Tags for consistency inside BCL APIs.
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? tags)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+### ActivityLink
+
+```C#
+    // Already exposed APIs
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+## Automatic Trace Id  generation in case of null parent
+
+Sampler like probabilistic sampler can depend on the trace id for the sampling decision. Usually before creating any Activity using ActivitySource, the `ActivityListener.GetRequestedDataUsingContext` callback will get called which usually get implemented by the samplers. The parent context is sent as a parameter to this callback and the sampler gets the trace id from this parent context to perform the sampling action. When there is no parent, we send the default context which is nulling the trace id. The samplers depending on trace id are not going to be able to do the sampling at that time. The samplers even cannot generate and use a trace id because it cannot inform the Activity creation process to use the generated trace Id when creating the Activity object.
```

Comment:
Can you add about the proposal @noahfalk said yesterday about `ActivityListener.GetRequestedDataUsingParentId` - This becoming an optional thing, and .NET will provide a default implementation where it'll try to generate `ActivityContext` from the string parentid, and invoke the other callback `ActivityListener.GetRequestedDataUsingContext`.
if ActivityContext cannot be extracted from parentid, then the default implementation could return .None.

Reply:
>  This becoming an optional thing, and .NET will provide a default implementation where it'll try to generate ActivityContext from the string parentid, and invoke the other callback ActivityListener.GetRequestedDataUsingContext.

I am not sure that what @noahfalk said. I understood that will be the listener's job to do I think and not the .NET. I may be mistaken but @noahfalk can clarify what he said.

Reply:
@noahfalk Can you confirm this? I believe my comment above is what you said, but please do confirm/correct me :)

Reply:
Yep! @cijothomas captured my suggestion accurately : )

Reply:
> if ActivityContext cannot be extracted from parentid, then the default implementation could return .None.

What do you mean by default implementation here?

Reply:
I meant .NET ActivitySource/Listener will provide a default implementation for `GetRequestedDataUsingParentId`. OpenTelemetry (or others) may optionally override it, if they are not happy with the default.

The default implementation (part of DiagnoticSource package) is something like this:

```
GetRequestedDataUsingParentId = (ref ActivityCreationOptions<string> options) => 
{
var parentid = options.Parent;
if (IsW3C(parentID))
  {
    var activityContext = Extract(parentId);
    call the GetRequestedDataUsingContext()
  }
else
{
   ActivityDataRequest.None;
}
},
```

Reply:
Shouldn't this be handled on the level of the `ActivitySource.StartActivity(string name, ActivityKind kind, string parentId, IEnumerable<KeyValuePair<string, string?>>? tags = null, IEnumerable<ActivityLink>? links = null, DateTimeOffset startTime = default)` instead? 

I mean the implementation of this API will try to extract the context from the parent Id and will call `GetRequestedDataUsingContext`. If it couldn't extract the context, will continue calling `GetRequestedDataUsingParentId`. If the listener cares about `GetRequestedDataUsingParentId`, can implement it. If the listener doesn't care, it should leave it null in the listener (which effectively is the same as returning None). Doing your suggestion will make GetRequestedDataUsingParentId is active for all listeners which all of them will pay the cost of this extra call while it is not really needed.

Reply:
You are right. I focused too much on the GetRequestedDataUsingParentId callback!

Reply:
Thanks, @cijothomas. It is always good to hear other ideas and compare and choose the best. your feedback is really valuable.

Reply:
I have updated the doc according to this discussion.